Sources/Tracking/CustomerIO.swift

```diff
@@ -238,6 +246,59 @@ public class CustomerIO: CustomerIOInstance {
             )
     }
 
+    // Swizzled method for APN device token.
+    @objc
+    func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
+        let token = String(apnDeviceToken: deviceToken)
+        Self.shared.registerDeviceToken(token)
+    }
+
+    // Implement method swizzling for APN device token. Swizzling will ensure that the
+    // device token gets registered without needing the developer to add call respective
+    // Customer.io SDK methods to register the device.
+    @available(iOSApplicationExtension, unavailable)
+    func swizzleDidReceiveRemoteNotification() {
+        DispatchQueue.main.async {
+            let appDelegate = UIApplication.shared.delegate
+            let appDelegateClass: AnyClass? = object_getClass(appDelegate)
+            let originalSelector = #selector(UIApplicationDelegate.application(_:didRegisterForRemoteNotificationsWithDeviceToken:))
+            let swizzledSelector = #selector(self.application(_:didRegisterForRemoteNotificationsWithDeviceToken:))
+            guard let swizzledMethod = class_getInstanceMethod(CustomerIO.self, swizzledSelector) else {
+                return
+            }
+            if let originalMethod = class_getInstanceMethod(appDelegateClass, originalSelector) {
+                method_exchangeImplementations(originalMethod, swizzledMethod)
+            }
+        }
+    }
+
+    // Swizzled method for FCM Token. This method would not be called if the app is using APNS.
+    @objc
+    func messaging(_ messaging: Messaging, didReceiveRegistrationToken fcmToken: String?) {
+        if let token = fcmToken {
+            Self.shared.registerDeviceToken(token)
+        }
+    }
+
+    // Implement method swizzling for FCM token. Swizzling the FCM token method will ensure
+    // that the token is registered with CIO even when the developer doesn't call the respective
+    // CIO method after token generation.
+    @available(iOSApplicationExtension, unavailable)
+    func swizzleDidReceiveFCMRegistrationToken() {
+        DispatchQueue.main.async {
+            let messagingDelegate = Messaging.messaging().delegate
```

Comment:
should we probably keep a weak reference for messaging delegates to avoid any possible memory leakages?

```
weak var messagingDelegate
```